test/widgets/profile_test.dart

```diff
@@ -0,0 +1,269 @@
+          mkCustomProfileField(0, CustomProfileFieldType.shortText),
+          mkCustomProfileField(1, CustomProfileFieldType.longText),
+          mkCustomProfileField(2, CustomProfileFieldType.choice, fieldData: '{"x": {"text": "choiceValue", "order": "1"}}'),
+          mkCustomProfileField(3, CustomProfileFieldType.date),
+          mkCustomProfileField(4, CustomProfileFieldType.link),
+          mkCustomProfileField(5, CustomProfileFieldType.user),
+          mkCustomProfileField(6, CustomProfileFieldType.externalAccount, fieldData: '{"subtype": "external1"}'),
+          mkCustomProfileField(7, CustomProfileFieldType.pronouns),
+        ], realmDefaultExternalAccounts: {
+          'external1': RealmDefaultExternalAccount(
+            name: 'external1',
+            text: '',
+            hint: '',
+            urlPattern: 'https://example/%(username)s')});
+
+      final testCases = [
+        (find.text('field0'), find.text('shortTextValue'), CustomProfileFieldType.shortText),
+        (find.text('field1'), find.text('longTextValue'), CustomProfileFieldType.longText),
+        (find.text('field2'), find.text('choiceValue'), CustomProfileFieldType.choice),
+        (find.text('field3'), find.text('dateValue'), CustomProfileFieldType.date),
+        (find.text('field4'), find.text('http://example/linkValue'), CustomProfileFieldType.link),
+        (find.text('field5'), find.byType(Avatar), CustomProfileFieldType.user),
+        (find.text('field6'), find.text('http://example/externalValue'), CustomProfileFieldType.externalAccount),
+        (find.text('field7'), find.text('pronounsValue'), CustomProfileFieldType.pronouns),
+      ];
+      final profileDataTable = find.byType(Column).last;
+      testCases.map((testCase) {
+        Finder labelFinder = testCase.$1;
+        Finder fieldFinder = testCase.$2;
+        CustomProfileFieldType testCaseType = testCase.$3;
+        check(
+          because: 'find label for $testCaseType',
+          find.descendant(of: profileDataTable, matching: labelFinder).evaluate().length
+        ).equals(1);
+        check(
+          because: 'find field for $testCaseType',
+          find.descendant(of: profileDataTable, matching: fieldFinder).evaluate().length
+        ).equals(1);
+      });
+    });
+
+    testWidgets('page builds; error page shows up if data is missing', (WidgetTester tester) async {
+      await setupPage(tester, pageUserId: eg.selfUser.userId + 1989);
+      check(because: 'find no user avatar', find.byType(Avatar).evaluate()).isEmpty();
+      check(because: 'find error icon', find.byIcon(Icons.error).evaluate()).isNotEmpty();
+    });
+
+    testWidgets('page builds; link type will navigate', (WidgetTester tester) async {
+      const testUrl = 'http://example/url';
+      final user = eg.user(userId: 1, profileData: {
+        0: ProfileFieldUserData(value: testUrl),
+      });
+
+      await setupPage(tester,
+        users: [user],
+        pageUserId: user.userId,
```

Comment:
optional nit: can simplify several of these by having setupPage take something like `User pageUser, List<User> otherUsers = []`. Then this can just say `pageUser: user`, which also lets it conveniently inline `user`; and even the call sites that do want to pass additional users also get to avoid referring twice to `user`.

Reply:
Ignoring this for now, the test case for the error page needs to pass in an invalid `pageUserId` (one that doesn't exist in the store) and I don't believe your proposed `setupPage` will allow that - unless I added in an extra flag, which complicates the function in another way, or had the error page test duplicate the setup.

Reply:
Hmm yeah, I see. That'd require some other direction of refactoring, then.

> or had the error page test duplicate the setup.

One arrangement I've found helpful in some test files is to have the setup steps broken up into a couple of functions… looks like `test/model/message_list_test.dart` is the example I'm most thinking of:
```dart
  /// Initialize [model] and the rest of the test state.
  void prepare({Narrow narrow = const AllMessagesNarrow()}) {
// …

  /// Perform the initial message fetch for [model].
  ///
  /// The test case must have already called [prepare] to initialize the state.
  Future<void> prepareMessages({
    required bool foundOldest,
    required List<Message> messages,
  }) async {
```

Then many test cases want both, but because the parameters are disjoint, it's not much added verbosity even then. And some test cases want only `prepare`, so they benefit from the separation.

Anyway, definitely no need to do that for this PR.